docs/development/communication.md

```diff
@@ -1,11 +1,147 @@
+
+This chapter will give an overview of existing and popular protocols.
+
+## Requirements
+
+To be able to compare protocols, some specifications should be postulated. These are arbitrary based on the needs identified for Libre Solar devices and could be changed e.g. for automotive needs.
+
+### Lower layers
+
+1. **Reliability:** Paket loss, ordering of pakets for multiframe messages etc. should be handled. The connection should be somewhat resistant to noise since devices like the charge-controller use large conductances and high currents.
+
+2. **Masterless:** Bidirectional communication and broadcasting should be possible. Protocols like [I²C](https://de.wikipedia.org/wiki/I%C2%B2C) or [SPI](https://de.wikipedia.org/wiki/Serial_Peripheral_Interface), which are often used with Arduinos, are not suitable. They also have quite low maximum cable length.
+
+3. **Speed:** To transmit larger amounts of logging data or even firmware updates, the speed should be considered.
+
+### Application layer
+
+1. **Schema-less:** The data structure should be discoverable and self-explaining. This means the client does not need to know the schema of the devices data, e.g. variable types and units should be delivered with the data, executable functions should can be identified etc.
+
+2. **Flexible:** The protocol should be independent of those protocols used in lower layers. It is ok to assume data integrity completeness and correct order, which should be checked by lower layer protocols.
+
+3. **Stateless:** Each request should be atomic regarding the information needed to process it. No session handling is done.
+
+4. **Compact:** Overhead like headers or data representation should be compact. Especially with IoT, lower layer protocols and connections are not very powerful, so low data rates should be expected and addressed.
+
+## Lower layers
+
+### CAN
+
+The Controller Area Network (CAN) was developed by the automotive industry to connect the different systems within a vehicle. It is a multimaster, electrical noise tolerant two wire bus system. The logical level on the bus is achieved with a differential approach, allowing for cable distances of more than 100m. Each end of the bus has to be terminated using a termination resistor [1]. Differential logic means that the voltage difference between the two wires indicate wether a "high" (logic 1) or "low" (logic 0) is written. In CAN, these states are called "recessive" and "dominant" respectively. The voltage difference depends on the used standard, in 5V systems it is usualy around 2V.
+
+The CAN Standard itself defines only the physical and data-link layer (layer 1 and 2 in OSI) with a maximum size of 8 bytes. In order to send larger messages, a protocol like [ISO 15765-2 - ISO-TP](https://en.wikipedia.org/wiki/ISO_15765-2) can be used.
+
+CAN offers a transmission speed up to 5 MBits/s, which greatly decreases the possible wire length. With a speed of 500 KBits/s, wires up to 100m can be used.
```

Comment:
https://en.wikipedia.org/wiki/Data-rate_units
```suggestion
CAN offers a transmission speed up to 5 Mbit/s, which greatly decreases the possible wire length. With a speed of 500 kbit/s, wires up to 100m can be used.
```

Reply:
Why "kilo" lower case and "Mega" uppper case?

Reply:
Because that's the way it is... see also the linked Wikipedia article. Lower-case m stands for milli (which does not make sense for bits, though).